imports/server/publishJoinedQuery.ts

```diff
@@ -247,46 +247,47 @@ const addObservers = (
   });
 };
 
-export default class JoinPublisher<T extends { _id: string }> {
-  watcher: Meteor.LiveQueryHandle;
-
-  observers: Map<string, RefCountedJoinedObjectObserverMap<any>>;
-
-  constructor(
-    sub: Subscription,
-    spec: PublishSpec<T>,
-    query: Mongo.Selector<T>,
-  ) {
-    validateSpec(spec);
-
-    this.observers = new Map<string, RefCountedJoinedObjectObserverMap<any>>();
-    addObservers(sub, spec, this.observers);
-
-    const { model } = spec;
-    const { _name: name } = model;
-
-    const observer = this.observers.get(name)!;
-
-    this.watcher = model.find(query, {
-      fields: { _id: 1 },
-    }).observeChanges({
-      added: (id) => {
-        observer.incref(id);
-      },
-      removed: (id) => {
-        if (spec.lingerTime !== undefined) {
-          Meteor.setTimeout(() => { observer.decref(id); }, spec.lingerTime);
-        } else {
-          observer.decref(id);
-        }
-      },
-    });
+export interface JoinedQueryHandle {
+  stop: () => void;
+}
 
-    sub.ready();
-  }
+export default function publishJoinedQuery<T extends { _id: string }>(
+  sub: Subscription,
+  spec: PublishSpec<T>,
+  query: Mongo.Selector<T>,
+): JoinedQueryHandle {
+  validateSpec(spec);
+
+  const observers = new Map<string, RefCountedJoinedObjectObserverMap<any>>();
+  addObservers(sub, spec, observers);
+
+  const { model } = spec;
+  const { _name: name } = model;
+
+  const observer = observers.get(name)!;
+  let watcher: Meteor.LiveQueryHandle;
+
+  const stop = () => {
+    watcher.stop();
+    observers.forEach((v) => v.shutdown());
+  };
+
+  watcher = model.find(query, {
+    fields: { _id: 1 },
+  }).observeChanges({
+    added: (id) => {
+      observer.incref(id);
+    },
+    removed: (id) => {
+      if (spec.lingerTime !== undefined) {
+        Meteor.setTimeout(() => { observer.decref(id); }, spec.lingerTime);
+      } else {
+        observer.decref(id);
+      }
+    },
+  });
+  sub.onStop(() => stop());
```

Comment:
Since this will now be called on subscription shutdown, as well as manually, that means that when we call `stop` manually, the sub infrastructure will also call it a second time when the client unsubs.  We should probably make `stop()` explicitly tolerate being called multiple times (and do nothing after the first invocation).

Reply:
Now that `pendingGuessesForSelf` doesn't need a handle to the shutdown function, we can actually ensure that it only gets called once, so I think we can side-step this specific issue.